source/fuzz/transformation_replace_parameter_with_global.cpp

```diff
@@ -0,0 +1,248 @@
+
+        // Insert an OpStore before the OpFunctionCall. +1 since the first
+        // operand of OpFunctionCall is an id of the function.
+        inst->InsertBefore(MakeUnique<opt::Instruction>(
+            ir_context, SpvOpStore, 0, 0,
+            opt::Instruction::OperandList{
+                {SPV_OPERAND_TYPE_ID, {message_.global_variable_fresh_id()}},
+                {SPV_OPERAND_TYPE_ID,
+                 {inst->GetSingleWordInOperand(parameter_index + 1)}}}));
+
+        // +1 since the first operand of OpFunctionCall is an id of the
+        // function.
+        inst->RemoveInOperand(parameter_index + 1);
+      });
+
+  // Remove the parameter from the function.
+  function->RemoveParameter(message_.parameter_id());
+
+  // Update function's type.
+  auto* old_function_type = fuzzerutil::GetFunctionType(ir_context, function);
+  assert(old_function_type && "Function has invalid type");
+
+  // Preemptively add function's return type id.
+  std::vector<uint32_t> type_ids = {
+      old_function_type->GetSingleWordInOperand(0)};
+
+  // +1 and -1 since the first operand is the return type id.
+  for (uint32_t i = 1; i < old_function_type->NumInOperands(); ++i) {
+    if (i - 1 != parameter_index) {
+      type_ids.push_back(old_function_type->GetSingleWordInOperand(i));
+    }
+  }
+
+  if (ir_context->get_def_use_mgr()->NumUsers(old_function_type) == 1) {
+    // Change the old type in place. +1 since the first operand is the result
+    // type id of the function.
+    old_function_type->RemoveInOperand(parameter_index + 1);
+  } else {
+    // Find an existing or create a new function type.
+    function->DefInst().SetInOperand(
+        1, {fuzzerutil::FindOrCreateFunctionType(
+               ir_context, message_.function_type_fresh_id(), type_ids)});
+  }
+
+  // Make sure our changes are analyzed
+  ir_context->InvalidateAnalysesExceptFor(
+      opt::IRContext::Analysis::kAnalysisNone);
+}
+
+protobufs::Transformation TransformationReplaceParameterWithGlobal::ToMessage()
+    const {
+  protobufs::Transformation result;
+  *result.mutable_replace_parameter_with_global() = message_;
+  return result;
+}
+
+bool TransformationReplaceParameterWithGlobal::CanReplaceFunctionParameterType(
+    const opt::analysis::Type& type) {
+  // TODO(https://github.com/KhronosGroup/SPIRV-Tools/issues/3403):
+  //  Think about other type instructions we can add here.
+  switch (type.kind()) {
+    case opt::analysis::Type::kBool:
+    case opt::analysis::Type::kInteger:
+    case opt::analysis::Type::kFloat:
+      return true;
+    case opt::analysis::Type::kArray:
+      return CanReplaceFunctionParameterType(*type.AsArray()->element_type());
+    case opt::analysis::Type::kMatrix:
+      return CanReplaceFunctionParameterType(*type.AsMatrix()->element_type());
+    case opt::analysis::Type::kVector:
+      return CanReplaceFunctionParameterType(*type.AsVector()->element_type());
```

Comment:
Matrices and vectors can only contain things that you'll definitely be able to replace, so I think you could include them in the first case when you return true.  E.g., you can't have a OpRuntimeArray in a vector.

Reply:
Fixed in febf42e.